File: 0037-sudoku-solver/0037-sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        if(solve(board)) return;
    }

    private boolean solve(char[][] board) {
        int row = -1;
        int col = -1;

        boolean foundEmpty = true;

        for(int i = 0; i < 9; i++) {
            for(int j = 0; j < 9; j++) {
                if(board[i][j] == '.') {
                    row = i;
                    col = j;
                    foundEmpty = false;
                    break;
                }
            }
            if(!foundEmpty) break;
        }

        if(foundEmpty) return true;

        for(char num = '1'; num <= '9'; num++) {
            if(isValid(num, board, row, col)) {
                board[row][col] = num;
                if(solve(board)) return true;
                board[row][col] = '.';
            }
        }

        return false;
        
    }

    private boolean isValid(char num, char[][] board, int row, int col) {
        // check row
        for(int i = 0; i < 9; i++) {
            if(board[row][i] == num) return false;
            if(board[i][col] == num) return false;
        }

        // check grid
        int x = row - row % 3;
        int y = col - col % 3;

        for(int i = 0; i < 3; i++) {
            for(int j = 0; j < 3; j++) {
                if(board[x+i][y+j] == num) return false;
            }
        }
        return true;

    }
}

File: 0037-sudoku-solver/0037-sudoku-solver.cpp
class Solution {
public:
    bool solve(vector<vector<char>>& board, int row, int col) {
        if (row == 9) return true;
        
        if (col == 9)
            return solve(board, row + 1, 0);
        
        if (board[row][col] != '.')
            return solve(board, row, col + 1);
        
        for (char num = '1'; num <= '9'; num++) {
            bool valid = true;
            
            // Check row
            for (int j = 0; j < 9; j++) {
                if (board[row][j] == num) {
                    valid = false;
                    break;
                }
            }
            
            // Check column
            for (int i = 0; i < 9 && valid; i++) {
                if (board[i][col] == num) {
                    valid = false;
                    break;
                }
            }
            
            // Check 3x3 box
            int sr = (row / 3) * 3;
            int sc = (col / 3) * 3;
            
            for (int i = sr; i < sr + 3 && valid; i++) {
                for (int j = sc; j < sc + 3; j++) {
                    if (board[i][j] == num) {
                        valid = false;
                        break;
                    }
                }
            }
            
            if (valid) {
                board[row][col] = num;
                
                if (solve(board, row, col + 1))
                    return true;
                
                board[row][col] = '.';
            }
        }
        
        return false;
    }

    void solveSudoku(vector<vector<char>>& board) {
        solve(board, 0, 0);
    }
};